Lib/functools.py

```diff
@@ -1037,36 +1034,45 @@ def register(self, cls, method=None):
         return self.dispatcher.register(cls, func=method)
 
     def __get__(self, obj, cls=None):
-        if self._method_cache is not None:
-            try:
-                _method = self._method_cache[obj]
-            except TypeError:
-                self._method_cache = None
-            except KeyError:
-                pass
-            else:
-                return _method
+        return _singledispatchmethod_get(self, obj, cls)
 
-        dispatch = self.dispatcher.dispatch
-        funcname = getattr(self.func, '__name__', 'singledispatchmethod method')
-        def _method(*args, **kwargs):
-            if not args:
-                raise TypeError(f'{funcname} requires at least '
-                                '1 positional argument')
-            return dispatch(args[0].__class__).__get__(obj, cls)(*args, **kwargs)
+    @property
+    def __isabstractmethod__(self):
+        return getattr(self.func, '__isabstractmethod__', False)
 
-        _method.__isabstractmethod__ = self.__isabstractmethod__
-        _method.register = self.register
-        update_wrapper(_method, self.func)
 
-        if self._method_cache is not None:
-            self._method_cache[obj] = _method
+class _singledispatchmethod_get:
+    def __init__(self, unbound, obj, cls):
+        self._unbound = unbound
+        self._dispatch = unbound.dispatcher.dispatch
+        self._obj = obj
+        self._cls = cls
 
-        return _method
+    def __call__(self, /, *args, **kwargs):
+        if not args:
+            funcname = getattr(self._unbound.func, '__name__',
+                               'singledispatchmethod method')
+            raise TypeError(f'{funcname} requires at least '
+                            '1 positional argument')
+        return self._dispatch(args[0].__class__).__get__(self._obj, self._cls)(*args, **kwargs)
+
+    def __getattr__(self, name):
+        if name not in {'__name__', '__qualname__', '__isabstractmethod__',
+                        '__annotations__', '__type_params__'}:
+            raise AttributeError
+        return getattr(self._unbound.func, name)
 
     @property
-    def __isabstractmethod__(self):
-        return getattr(self.func, '__isabstractmethod__', False)
+    def __module__(self):
```

Comment:
This is giving some problems in my interactive console. A minimal reproducer:
```
from functools import singledispatchmethod, singledispatch
from IPython.lib.pretty import pretty

class A:
    def __init__(self, value):
        self.value = value
        
    @singledispatchmethod
    def dp(self, x):
        return id(self)
    
a=obj=A(4)

pretty(a.dp) # fails

_singledispatchmethod_get.__module__ # this is now a property, not a string

```

Reply:
Yes, there is a problem with type attributes which we want to define for instances (`__name__`, `__qualname__`, `__doc__`). If we define a property, it conflicts with the type attribute.

Reply:
This can be resolved by defining `__getattribute__` instead of `__getattr__`. But this adds such large overhead, that it outweigh the optimization gain.

So I added setting just two instance attributes `__module__` and `__doc__` in the constructor. This adds some overhead, but not so large as in the original code. I hope that more satisfying solution will be found in future, but this is a complex issue.